File: java/anagram/src/main/java/Anagram.java
import java.util.Arrays;
import java.util.List;

class Anagram {

    private String word;

    public Anagram(String word) {
        this.word = word.toLowerCase();
    }

    public List<String> match(List<String> candidates) {
        return candidates.stream().filter(this::isAnagram).toList();
    }

    private boolean isAnagram(String candidate) {
        return !candidate.toLowerCase().equals(this.word) 
                    && Arrays.equals(
                            candidate.toLowerCase().chars().sorted().toArray(),
                            this.word.chars().sorted().toArray()
                        );
    }
}

File: javascript/anagram/anagram.js
export const findAnagrams = (word, candidates) => 
    candidates
        .filter(candidate => candidate.toLowerCase() !== word.toLowerCase())
        .filter(candidate => isAnagram(word.toLowerCase(), candidate.toLowerCase()));

const isAnagram = (word, candidate) => 
    [...word].toSorted().join('') === [...candidate].toSorted().join('');

